Show star rating / review summary in review listing

kumar303 opened the pull request:
Fixes https://github.com/mozilla/addons-frontend/issues/3919

kumar303 commented on src/ui/components/Rating/index.js:
```diff
@@ -85,20 +94,35 @@ export class RatingBase extends React.Component<InternalProps> {
         }),
-        id: `Rating-rating-${thisRating}`,
         key: `rating-${thisRating}`,
+        onMouseEnter: () => this.onHoverStar(thisRating),
```
I couldn't get a pure CSS hover to work due to the `grid-gap` which causes a flicker. I couldn't think of an alternative that didn't involve implementing a custom gap which is a pitfall of doom. Switching to grid fixes so many other layout issues that I think it's an acceptable trade-off.

kumar303 commented on src/amo/components/AddonReview/styles.scss:
```diff
@@ -10,6 +10,12 @@
 
   .Rating {
     margin: 6px 0 12px;
+    padding: 0 10px;
```
The values in this file are inconsistent but I was just preventing regressions. This component is going away soon.

kumar303 commented on src/ui/components/Rating/index.js:
```diff
@@ -28,14 +28,23 @@ type InternalProps = {|
 
+  constructor(props: InternalProps) {
+    super(props);
+    this.state = { hoveringOverStar: null };
```
I chose `setState` here because I wanted the UI to be snappy when hovering. Switching to `setUIState()` also burdens the caller with having to generate an ID. I dunno, I'm still thinking on it.

rebmullin commented on src/ui/components/Rating/index.js:
```diff
@@ -85,20 +94,35 @@ export class RatingBase extends React.Component<InternalProps> {
     return [1, 2, 3, 4, 5].map((thisRating) => {
+      let isSelected = thisRating - rating <= 0.25;
+      if (this.state.hoveringOverStar !== null) {
+        isSelected = thisRating <= this.state.hoveringOverStar;
```
can we extract out hoveringOverStar since it's used a few times?

rebmullin commented on src/amo/components/RatingManager/styles.scss:
```diff
@@ -30,4 +23,10 @@ $button-height: $rating-l-star-size + $rating-l-star-gutter -
+
+  .Rating {
+    @include respond-to(large) {
+      margin: 12px;
```
maybe we could keep this margin at smaller viewports too? At least on top:

![Alt text](https://monosnap.com/image/c3Ux0Pyuh9y2Wutn7yimxpjBCQj7rE.png)

what do you think?

kumar303 replied:
Good idea, thanks.

bobsilverberg commented on src/ui/components/Rating/index.js:
```diff
@@ -85,20 +94,35 @@ export class RatingBase extends React.Component<InternalProps> {
           'Rating-half-star':
             thisRating - rating > 0.25 && thisRating - rating <= 0.75,
         }),
-        id: `Rating-rating-${thisRating}`,
```
Why don't we need an `id`? Is it because we are rendering an html element and not a react component?

kumar303 replied:
The id was just used for testing. It doesn't really make sense to create an ID like this because on a page like the review listing, there could be many versions of `Rating` with `thisRating==4`.

kumar303 replied:
Oh, wait, you mean for the JSX? Ah, yes, it does need an `id`. Good catch! The linter didn't catch this because I was setting it in a props object and using a spread for the JSX constructor.

kumar303 replied:
Ignore that last comment. Heh. I was confusing `id` with `key`. Each array item already has a `key` so this will keep DOM mutations to a minimum.

bobsilverberg replied:
Ha, I was confusing `id` with `key` too!

kumar303 commented:
> I did have 1 question about the location of the stars on the reviews page - the mock shows it up top. Is that just an old mock?

@rebmullin thanks for the review. The mock you referenced is up to date. Are you saying that the star rating graph does not appear at the top of the page? Can you send me a URL and screenshot? It appears at the top for me: https://screenshots.firefox.com/m65oS801QVj6sXne/localhost

bobsilverberg commented on tests/unit/amo/components/TestAddonReviewList.js:
```diff
@@ -517,13 +525,11 @@ describe(__filename, () => {
 
-      expect(h3.render().text()).toEqual('by Hayley');
-      expect(h3.render().find('a')).toHaveLength(0);
+      expect(h3.text()).toEqual('by Hayley');
```
Nit: Ideally `Haley` would be in a `const` rather than hardcoded twice.

bobsilverberg commented on tests/unit/amo/components/TestAddonReviewList.js:
```diff
@@ -540,13 +546,11 @@ describe(__filename, () => {
-
-      expect(h3.render().text()).toEqual('by Chantal, Leroy');
-      expect(h3.render().find('a')).toHaveLength(2);
+      expect(h3.text()).toEqual('by Chantal, Leroy');
```
Nit: Ideally `Chantal` and `Leroy` would be in `const`s rather than hardcoded twice.

bobsilverberg commented on tests/unit/amo/components/TestAddonReviewList.js:
```diff
@@ -540,13 +546,11 @@ describe(__filename, () => {
-      expect(h3.render().text()).toEqual('by Chantal, Leroy');
-      expect(h3.render().find('a')).toHaveLength(2);
+      expect(h3.text()).toEqual('by Chantal, Leroy');
+      expect(h3.find('a')).toHaveLength(2);
```
Should the test also be checking that the `href` of the `a` tags are the expected values?

bobsilverberg commented on tests/unit/amo/components/TestAddonReviewList.js:
```diff
@@ -640,76 +644,100 @@ describe(__filename, () => {
 
-      sinon.assert.calledWith(dispatchSpy, clearAddonReviews({ addonSlug }));
+      expect(ratingsByStar).toHaveProp('addon');
+      expect(ratingsByStar.prop('addon')).toMatchObject({ id: addon.id });
```
Why not just `expect(ratingsByStar).toHaveProp('addon', addon);`?

kumar303 replied:
Because the test doesn't access to the internal add-on object. I added a comment to make it less confusing.

bobsilverberg commented on tests/unit/amo/components/TestAddonReviewList.js:
```diff
@@ -640,76 +644,100 @@ describe(__filename, () => {
+        ...fakeAddon,
+        ratings: {
+          ...fakeAddon.ratings,
+          average: 4.6667,
```
Nit: Would it be better to use a `const` for this and in the assertion?

kumar303 replied:
Since our test files don't use Flow, I think it's a bit safer to hard code conversions like this. The alternative would be to re-implement the `.toFixed(1)` call but that kind of thing can lead to bitrot without static typing.

bobsilverberg commented on tests/unit/amo/components/TestAddonReviewList.js:
```diff
@@ -640,76 +644,100 @@ describe(__filename, () => {
+        ...fakeAddon,
+        ratings: {
+          ...fakeAddon.ratings,
+          average: 4.0,
```
Nit: Would it be better to use a `const` for this and in the assertion?

kumar303 replied:
Same reason as above.

bobsilverberg commented on tests/unit/ui/components/TestRating.js:
```diff
@@ -227,6 +227,51 @@ describe(__filename, () => {
+    const rating = 3;
+    const hoverStar = getStar({ root, rating });
+    hoverStar.simulate('mouseEnter', createFakeEvent());
+
```
Would it be useful to have a check assertion in here that the star has the `Rating-selected-star` class, to avoid a possible false positive for the next assertion?

kumar303 replied:
I don't think it would be useful because that's a duplication of the previous test. The false positive wouldn't be an issue since the logic is already covered -- at least one test would fail.

bobsilverberg commented on tests/unit/ui/components/TestRating.js:
```diff
@@ -238,14 +283,26 @@ describe(__filename, () => {
+      const hoverStar = getStar({ root, rating });
+      hoverStar.simulate('mouseEnter', createFakeEvent());
+
+      expect(getStar({ root, rating })).not.toHaveClassName(
```
This test is checking for a specific behaviour being absent, but the test says "does nothing when you hover over stars", so would it be any better to check that the classNames of the star are the same before and after simulating the hover, rather than checking for the absence of this specific class?

rebmullin commented:
> The mock you referenced is up to date. Are you saying that the star rating graph does not appear at the top of the page? Can you send me a URL and screenshot? It appears at the top for me: screenshots.firefox.com/m65oS801QVj6sXne/localhost

hey @kumar303, sorry for not being clearer. And actually what I am bringing up might be technically out of scope (of your PR anyway) ..but just to clarify, I was referring to the stars pointed to here:
![Alt text](https://monosnap.com/image/F0RdoomdG1zjrTnxYlLWCmdp5V66zx.png)

otherwise is looking good to me 👍

kumar303 commented:
@bobsilverberg 

> I've looked over this and I'm a bit confused about why so many changes were necessary to implement this for the review listing.

- The old implementation never worked quite right so lots of components that rendered ratings had to employ hacks to get things to work. These all needed updating and this makes up the bulk of the changes.
- Part of this patch was removing `AddonReviewList`'s ability to open an editor for the rating. When I removed this I simplified the logic and tests but that created a lot of churn in the diff.

> I'm also a bit surprised that we need a grid for the Rating component.

I am open to other implementation ideas but I think grid is a good fit.

Adding big rating stars to the top of the pre-existing `RatingsByStar` component seemed like a simple patch but the first problem I ran into was that `Rating` was implemented with fixed star sizes. It was not possible to match the mocks without adding a new custom size and this got me thinking that the current implementation didn't handle multiple device sizes well at all. 

I couldn't think of a way to make it more responsive with the flex-box implementation -- trying to account for a gutter and dealing with the quirks of percentage widths just didn't seem like a good use of time, especially considering that grid is designed specifically for this type of layout.

Once I got through the grid implementation, I remembered that hovering over stars was achieved with a sibling selector and that won't work over the grid gaps. D'oh. It's unfortunate that I had to resort to hover events but I still think the pros of the grid shine through. 

The stars are now much easier to place inside of content (notice that all negative margin hacks are gone!) and the layout looks a lot better when you resize the screen.

kumar303 commented on src/ui/components/Rating/styles.scss:
```diff
@@ -1,41 +1,52 @@
+}
+
+.Rating-star {
+  background: url('./img/closed-star-dim-gray.svg') center/contain no-repeat;
```
They could be! I thought about this too while working on it but didn't want to explode the patch any further. I filed this to use an inline SVG which I think would solve it nicely. ﻿﻿https://github.com/mozilla/addons-frontend/issues/5925

kumar303 commented on tests/unit/amo/components/TestAddonReviewList.js:
```diff
@@ -100,15 +96,27 @@ describe(__filename, () => {
+
+      expect(
+        header.find('.AddonReviewList-header-icon img').prop('src'),
+      ).toEqual(fallbackIcon);
```
Ah, thanks. I was moving fast trying to the get the old code working.
